### DISCUSSION TITLE
WPF-style binding not work

### DISCUSSION THREAD

**Discussion by wocow3** [AvaloniaUI/Avalonia#17786] [Q&A]
Our WPF project has a number of bindings used like this
```
  public class TestClass {
	  public int TestNumber { get; set; }
  };
  public class MainViewModel : ObservableObject
   {
	  private TestClass _testClass;
	  public TestClass TestClass
	  {
		  get => _testClass;
		  set => SetProperty(ref _testClass, value);
	  }
	  public ICommand TestCommand { get; }
  }
  public MainViewModel()
  {
	  TestClass = new TestClass();
	  TestCommand = new RelayCommand(
		  () => {
			  TestClass.TestNumber = TestClass.TestNumber + 1;
			  OnPropertyChanged("TestClass");
		  }
	  );
  }
```

```
<Button Command="{Binding TestCommand}">TestCommand</Button>
<TextBlock Text="{Binding TestClass.TestNumber}">
```
Bindings are notified by an explicit event invoke through 'TestClass' mid node
But this not work in Avalonia, how can we acheive that ?

**Comment by IanRawley** [comment]
TestClass doesn't implement INotifyPropertyChanged, so the change in its TestNumber property isn't passed on to anyone.  The manual raising of PropertyChanged for the TestClass property on the MainViewModel gets filtered out because TestClass hasn't actually changed.

To get this working in Avalonia TestClass will also have to implement INotifyPropertyChanged, and raise notifications when TestNumber changes.

**Reply by wocow3** [reply]
Thanks for your reply.
OnPropertyChanged does not check the value, but directly invoke event. WPF does not require all nodes of PropertyPath to implement INotifyPropertyChanged. Only one node is required to implement. 
It should be Avalonia's redesigned binding mechanism that makes this trick no longer work

**Reply by IanRawley** [reply]
Yes.  That's the big difference.  WPF automatically creates INotifyPropertyChanged wrappers around nodes that don't implement it themselves, while Avalonia does not.  So to get the example working in Avalonia you'll need to implement INotifyPropertyChanged at all levels, either directly or by creating your own wrappers if you can't change the model.
